### PR TITLE
docs: Expand speeding up training guide with acceleration methods

### LINKED ISSUE #4382: Populate "Speeding Up Training"
Currently, this section only mentions vLLM. We should have a small guide for other methods, like flash attention.
Ideally, to avoid repetition, we should have a very light example, and a link to the place in the doc where it's more extensively discussed, example vLLM pointing to vLLM integration guide

### PR DESCRIPTION
**Pull request by behroozazarkhalili**
## Summary

This PR expands the "Speeding Up Training" documentation to include multiple acceleration methods beyond vLLM, addressing issue #4382.

## Changes

- **Flash Attention 2**: Added section explaining optimized attention mechanism with minimal example using `model_init_kwargs`
- **PEFT Integration**: Added section on parameter-efficient fine-tuning with LoRA example
- **Liger Kernel**: Added section on Triton kernels for memory optimization
- **Gradient Checkpointing**: Added section on trading compute for memory
- **Mixed Precision Training**: Added section on bf16/fp16 training
- Updated introduction to reflect comprehensive coverage of acceleration methods

## Implementation Details

- Each section follows the established pattern: brief explanation + minimal code example + link to detailed docs
- All code examples verified against TRL source code and official examples
- Examples use consistent pattern with `...` for omitted parameters (matching existing vLLM section)
- Links point to existing comprehensive documentation to avoid redundancy

## Testing

All code patterns verified against:
- `tests/test_sft_trainer.py` for Flash Attention 2
- `examples/notebooks/sft_trl_lora_qlora.ipynb` for PEFT
- `docs/source/liger_kernel_integration.md` for Liger Kernel
- `examples/scripts/sft_gemma3.py` for gradient checkpointing and mixed precision

Resolves #4382

### CODE REVIEW

File: docs/source/speeding_up_training.md [hunk 1]
```diff
@@ -95,3 +94,80 @@ You can customize the server configuration by passing additional arguments. For
 
 </hfoption>
 </hfoptions>
+
+## Flash Attention 2 for faster attention computation
+
+Flash Attention 2 is an optimized implementation of the attention mechanism that can significantly speed up training while reducing memory usage. It's particularly effective for long sequences.
```
Comment (sergiopaniego) [inline docs/source/speeding_up_training.md]: I'd make this more general to different attention implementations and include a detail about all of them and additionally Kernels from the Hub (https://huggingface.co/docs/trl/kernels_hub) with the link to the guide.

File: docs/source/speeding_up_training.md [hunk 2]
```diff
@@ -95,3 +94,80 @@ You can customize the server configuration by passing additional arguments. For
...
+    ...,
+    model_init_kwargs={"attn_implementation": "flash_attention_2"}
+)
+```
+
+Flash Attention 2 works across all TRL trainers. For padding-free batching with Flash Attention, see [Reducing Memory Usage](reducing_memory_usage#padding-free).
+
+## PEFT for parameter-efficient training
+
+PEFT (Parameter-Efficient Fine-Tuning) methods like LoRA significantly reduce memory usage and training time by only training a small number of adapter parameters instead of the full model.
```
Comment (sergiopaniego) [inline docs/source/speeding_up_training.md]:
```suggestion
[PEFT](https://huggingface.co/docs/peft/index) (Parameter-Efficient Fine-Tuning) methods like LoRA significantly reduce memory usage and training time by only training a small number of adapter parameters instead of the full model.
```

File: docs/source/speeding_up_training.md [hunk 3]
```diff
@@ -95,3 +94,80 @@ You can customize the server configuration by passing additional arguments. For
...
+
+Liger Kernel is a collection of Triton kernels designed for LLM training that can increase throughput by 20% and reduce memory usage by 60%.
+
+```python
+from trl import DPOConfig
+
+training_args = DPOConfig(..., use_liger_kernel=True)
+```
+
+Liger Kernel is supported across multiple trainers (SFT, DPO, GRPO, KTO, GKD). For more information, see [Liger Kernel Integration](liger_kernel_integration).
```
Comment (sergiopaniego) [inline docs/source/speeding_up_training.md]: We could include the different trainers' examples in the code snippet.

File: docs/source/speeding_up_training.md [hunk 4]
```diff
@@ -95,3 +94,80 @@ You can customize the server configuration by passing additional arguments. For
...
+
+Mixed precision training using bf16 or fp16 can speed up training and reduce memory usage with minimal impact on model quality.
+
+```python
+from trl import SFTConfig
+
+training_args = SFTConfig(..., bf16=True)  # or fp16=True for older GPUs
+```
+
+Use `bf16=True` for Ampere GPUs (A100, RTX 30xx) or newer, and `fp16=True` for older GPUs. Mixed precision training is supported across all TRL trainers.
```
Comment (sergiopaniego) [inline docs/source/speeding_up_training.md:228]: Same ideas as in [here](https://github.com/huggingface/trl/pull/4427#discussion_r2487094041) :)

**Review by sergiopaniego — COMMENTED** [review]
Thanks for the improvement!!!

A few ideas :)

For the online methods in vLLM, you can add a link to the actual online methods in the taxonomy (https://moon-ci-docs.huggingface.co/docs/trl/pr_4428/en/index#online-methods). The code snippet for that part should include all the supported trainers :) 

Additionally, I'd add a link to the vLLM integration guide for more details and review that the different trainers' code snippets contain the same (currently the Online DPO is different)

**Review by sergiopaniego — APPROVED** [review]
thanks!
